x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilderTests.java

```diff
@@ -0,0 +1,287 @@
+
+    private static final int NUM_TOKENS = 10;
+
+    private static final Set<WeightedToken> WEIGHTED_TOKENS = Set.of(
+        new TextExpansionResults.WeightedToken("foo", .42f),
+        new TextExpansionResults.WeightedToken("bar", .05f),
+        new TextExpansionResults.WeightedToken("baz", .74f)
+    );
+
+    @Override
+    protected WeightedTokensQueryBuilder doCreateTestQueryBuilder() {
+        TokenPruningConfig tokenPruningConfig = randomBoolean()
+            ? new TokenPruningConfig(randomIntBetween(1, 100), randomFloat(), randomBoolean())
+            : null;
+
+        var builder = new WeightedTokensQueryBuilder(RANK_FEATURES_FIELD, WEIGHTED_TOKENS, tokenPruningConfig);
+        if (randomBoolean()) {
+            builder.boost((float) randomDoubleBetween(0.1, 10.0, true));
+        }
+        if (randomBoolean()) {
+            builder.queryName(randomAlphaOfLength(4));
+        }
+        return builder;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return List.of(MachineLearning.class, MapperExtrasPlugin.class);
+    }
+
+    @Override
+    protected boolean canSimulateMethod(Method method, Object[] args) throws NoSuchMethodException {
+        return method.equals(Client.class.getMethod("execute", ActionType.class, ActionRequest.class, ActionListener.class))
+            && (args[0] instanceof InferModelAction);
+    }
+
+    @Override
+    protected Object simulateMethod(Method method, Object[] args) {
+        InferModelAction.Request request = (InferModelAction.Request) args[1];
+        assertEquals(InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API, request.getInferenceTimeout());
+        assertEquals(TrainedModelPrefixStrings.PrefixType.SEARCH, request.getPrefixType());
+
+        // Randomisation of tokens cannot be used here as {@code #doAssertLuceneQuery}
+        // asserts that 2 rewritten queries are the same
+        var response = InferModelAction.Response.builder()
+            .setId(request.getId())
+            .addInferenceResults(List.of(new TextExpansionResults("foo", WEIGHTED_TOKENS.stream().toList(), randomBoolean())))
+            .build();
+        @SuppressWarnings("unchecked")  // We matched the method above.
+        ActionListener<InferModelAction.Response> listener = (ActionListener<InferModelAction.Response>) args[2];
+        listener.onResponse(response);
+        return null;
+    }
+
+    @Override
+    protected void initializeAdditionalMappings(MapperService mapperService) throws IOException {
+        mapperService.merge(
+            "_doc",
+            new CompressedXContent(Strings.toString(PutMappingRequest.simpleMapping(RANK_FEATURES_FIELD, "type=rank_features"))),
+            MapperService.MergeReason.MAPPING_UPDATE
+        );
+    }
+
+    /**
+     * Overridden to ensure that {@link SearchExecutionContext} has a non-null {@link IndexReader}
+     */
+    @Override
+    public void testToQuery() throws IOException {
```

Comment:
This is not right here, but I could use some guidance on how to correctly test this. 

Right now it's failing because the `TextExpansionQueryBuilder`'s `doToQuery()` method returns an `IllegalStateException` that this should have been written to another query type. I had to override this test, because if I don't this test fails because the `SearchExecutionContext` is null for this case. What's the best way to proceed for this type of situation? 

CC: @benwtrent @jimczi